File: Exámenes/2020/Febrero/Java/src/dataStructures/graph/FordFulkerson.java
/**
 * Student's name:
 *
 * Student's group:
 */

package dataStructures.graph;

import dataStructures.list.LinkedList;
import dataStructures.list.List;
import dataStructures.set.HashSet;
import dataStructures.set.Set;
import dataStructures.tuple.Tuple2;

public class FordFulkerson<V> {
    private WeightedDiGraph<V,Integer> g; // Initial graph 
    private List<WDiEdge<V,Integer>> sol; // List of edges representing maximal flow graph
    private V src; 			  // Source
    private V dst; 		  	  // Sink
	
    /**
     * Constructors and methods
     */

    public static <V> int maxFlowPath(List<WDiEdge<V,Integer>> path) {
        int max = 0;

        for (WDiEdge<V,Integer> arista : path) {
            if (arista.getWeight() > max) {
                max = arista.getWeight();
            }
        }

        return max;
    }

    public static <V> List<WDiEdge<V,Integer>> updateEdge(V x, V y, Integer p, List<WDiEdge<V,Integer>> edges) {
        List<WDiEdge<V,Integer>> res = new LinkedList<>();

        for (WDiEdge<V,Integer> arista : edges) {
            if (arista.getSrc() == x && arista.getDst() == y) {
                int peso = arista.getWeight();

                if (arista.getWeight() + p != 0) {
                    res.append(new WDiEdge<>(arista.getSrc(), peso + p, arista.getDst()));
                }

            }
        }

        return res;
    }

    public static <V> List<WDiEdge<V,Integer>> updateEdges(List<WDiEdge<V,Integer>> path, Integer p, List<WDiEdge<V,Integer>> edges) {
        List<WDiEdge<V,Integer>> res;
        WDiEdge<V,Integer> edge = edges.get(0);

        res = updateEdge(edge.getSrc(), edge.getDst(), p, path);

        int i = 1;

        while(i < edges.size()){
            res = updateEdge(edges.get(i).getSrc(), edges.get(i).getDst(), p, res);
        }

        return res;
    }

    public static <V> List<WDiEdge<V,Integer>> addFlow(V x, V y, Integer p, List<WDiEdge<V,Integer>> sol) {
        // TO DO
        List<WDiEdge<V,Integer>> res = new LinkedList<>();
        WDiEdge<V,Integer> arco = null;

        for (WDiEdge<V,Integer> arista : sol) {
            if (arista.getSrc() == x && arista.getDst() == y) {
                res = updateEdge(arista.getSrc(), arista.getDst(), p, sol);

            } else if (arista.getSrc() == y && arista.getDst() == x) {

            }
        }


        return null;
    }

    public static <V> List<WDiEdge<V,Integer>> addFlows(List<WDiEdge<V,Integer>> path, Integer p, List<WDiEdge<V,Integer>> sol) {
        // TO DO
        return null;
    }

    public FordFulkerson(WeightedDiGraph<V,Integer> g, V src, V dst) {
        // TO DO
    }

    public int maxFlow() {
        // TO DO
        return 0;
    }

    public int maxFlowMinCut(Set<V> set) {
        // TO DO
        return 0;
    }

    /**
     * Provided auxiliary methods
     */
    public List<WDiEdge<V, Integer>> getSolution() {
        return sol;
    }
	
    /**********************************************************************************
     * A partir de aquí SOLO para estudiantes a tiempo parcial sin evaluación continua.
     * ONLY for part time students.
     * ********************************************************************************/

    public static <V> boolean localEquilibrium(WeightedDiGraph<V,Integer> g, V src, V dst) {
        // TO DO
        return false;
    }
    public static <V,W> Tuple2<List<V>,List<V>> sourcesAndSinks(WeightedDiGraph<V,W> g) {
        // TO DO
        return null;
    }

    public static <V> void unifySourceAndSink(WeightedDiGraph<V,Integer> g, V newSrc, V newDst) {
        // TO DO
    }
}

File: Exámenes/2020/Febrero/Haskell/DataStructures/Graph/FordFulkerson.hs
-------------------------------------------------------------------------------
-- Ford-Fulkerson Algorithm. Maximal flow for a weighted directed graph.
--
-- Student's name:
-- Student's group:
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.FordFulkerson where

import Data.List  ((\\))
import DataStructures.Graph.WeightedDiGraph
import DataStructures.Graph.WeightedDiGraphBFT



maxFlowPath :: Path (WDiEdge a Integer) -> Integer
maxFlowPath [] = 0
maxFlowPath path = minimum [ w | (E f w t) <- path ]


updateEdge ::(Eq a) => a -> a -> Integer -> [WDiEdge a Integer] -> [WDiEdge a Integer]
updateEdge x y p []  = [(E x p y)]
updateEdge x y p ((E f w t):es) | pertenece && sumacero     = es
                                | pertenece && not sumacero = (E f (w+p) t) : es
                                | otherwise                 = (E f w t) : updateEdge x y p es
  where
    pertenece = x == f && y == t
    sumacero  = (p+w) == 0


updateEdges :: (Eq a) => Path (WDiEdge a Integer) -> Integer -> [WDiEdge a Integer] -> [WDiEdge a Integer]
updateEdges [] _ edges = edges
updateEdges ((E f w t):ps) p edges = updateEdges ps p (updateEdge f t p edges)


addFlow :: (Eq a) => a -> a -> Integer -> [WDiEdge a Integer] -> [WDiEdge a Integer]
addFlow x y p ((E f w t):ss) | perteneceX_Y = (E x (w+p) y):ss
                             | perteneceYPX = ss
                             | perteneceYwX = (E x (p-w) y):ss
                             | perteneceYWX = (E x (w-p) y):ss
                             | otherwise    = addFlow x y p ss
  where
    perteneceX_Y = x == f && y == t
    perteneceYPX = y == f && x == t && w == p
    perteneceYwX = y == f && x == t && w < p
    perteneceYWX = y == f && x == t && w > p


addFlows :: (Eq a) => Path (WDiEdge a Integer) -> Integer -> [WDiEdge a Integer] -> [WDiEdge a Integer]
addFlows [] _ sol = sol
addFlows ((E f w t):ps) p sol = addFlows ps p (addFlow f t p sol)


fordFulkerson :: (Ord a) => (WeightedDiGraph a Integer) -> a -> a -> [WDiEdge a Integer]
fordFulkerson g src dst | src == dst = [(E src 0 src)]
                        | otherwise  = undefined


maxFlow :: (Ord a) => [WDiEdge a Integer] -> a -> Integer
maxFlow sol src = sum [ w | (E f w t) <- sol , f == src ]


maxFlowMinCut :: (Ord a) => (WeightedDiGraph a Integer) -> a -> a -> [a] -> Integer
maxFlowMinCut g src dst set = undefined



-- A partir de aquí hasta el final
-- SOLO para alumnos a tiempo parcial
-- sin evaluación continua

localEquilibrium :: (Ord a) => WeightedDiGraph a Integer -> a -> a -> Bool
localEquilibrium = undefined

sourcesAndSinks :: (Eq a) => WeightedDiGraph a b -> ([a],[a])
sourcesAndSinks = undefined

unifySourceAndSink :: (Eq a) => WeightedDiGraph a Integer -> a -> a -> WeightedDiGraph a Integer
unifySourceAndSink = undefined
